Lexi DB panic

I was testing an unrelated change in #2769 with the tatanka demo and hit this panic, the first time and again and again.


```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x10de8fa82]

goroutine 200 [running]:
github.com/dgraph-io/badger/skl.(*Skiplist).IncrRef(...)
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/skl/skl.go:86
github.com/dgraph-io/badger.(*DB).getMemTables(0xc000452388)
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:493 +0xc2
github.com/dgraph-io/badger.(*DB).get(0xc000452388, {0xc000450d30, 0xb, 0xb})
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:525 +0x69
github.com/dgraph-io/badger.(*Txn).Get(0xc0002b7380, {0xc0004da020, 0x3, 0x3})
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/txn.go:407 +0x287
github.com/dgraph-io/badger.(*Sequence).Release.func1(0xc0002b7380)
        /Users/philemon/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:1131 +0x31
github.com/dgraph-io/badger.(*DB).Update(0xc000657380?, 0xc000574730)
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/txn.go:696 +0x74
github.com/dgraph-io/badger.(*Sequence).Release(0xc0001ed800)
        /Users/[redacted]/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:1130 +0x8b
decred.org/dcrdex/dex/lexi.(*DB).Connect.func1()
        /Users/[redacted]/decred/dcrdex/dex/lexi/lexi.go:84 +0x73
created by decred.org/dcrdex/dex/lexi.(*DB).Connect in goroutine 1
        /Users/[redacted]/decred/dcrdex/dex/lexi/lexi.go:81 +0x90
exit status 2

```

Log before the crash.
```
2025-03-24 19:24:02.828 [TRC] SRV[127.0.0.1:50952][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:24:02.847 [INF] SRV[127.0.0.1:50952]: Starting Tatanka node with peer ID 028d3efe5857a5831edf5f67c65583b9a36246d0227f004f7eae440a363627b48f
2025-03-24 19:24:02.847 [INF] SRV[127.0.0.1:50952]: Fiat rate source "CryptoCompare" is disabled...
2025-03-24 19:24:02.847 [INF] COMMS: Server listening on 127.0.0.1:50952
2025-03-24 19:24:02.847 [ERR] SRV[127.0.0.1:50952]: error connecting boot node with proto = 02ea332cf533d4e2043f16de1596cd3ebacaa626cec92cad3096143cc0beb8d7a6, config = {"url":"ws://127.0.0.1:50953","cert":null}
2025-03-24 19:24:03.829 [TRC] SRV[127.0.0.1:50953][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:24:03.841 [INF] SRV[127.0.0.1:50953]: Starting Tatanka node with peer ID 02ea332cf533d4e2043f16de1596cd3ebacaa626cec92cad3096143cc0beb8d7a6
2025-03-24 19:24:03.841 [INF] SRV[127.0.0.1:50953]: Fiat rate source "CryptoCompare" is disabled...
2025-03-24 19:24:03.842 [INF] COMMS: Server listening on 127.0.0.1:50953
2025-03-24 19:24:03.844 [DBG] COMMS: Starting websocket handler for 127.0.0.1:50957
2025-03-24 19:24:03.844 [TRC] COMMS: New websocket client 127.0.0.1
2025-03-24 19:24:03.844 [INF] SRV[127.0.0.1:50953]: Connected to boot node with peer ID 028d3efe5857a5831edf5f67c65583b9a36246d0227f004f7eae440a363627b48f, config {"url":"ws://127.0.0.1:50952","cert":null}
2025-03-24 19:24:03.844 [TRC] COMMS[WS]: Starting websocket messaging with peer 127.0.0.1
2025-03-24 19:24:03.845 [TRC] SRV[127.0.0.1:50952]: Tatanka node handling message. route = tatanka_connect, payload = {"id":[2,234,51,44,245,51,212,226,4,63,22,222,21,150,205,62,186,202,166,38,206,201,44,173,48,150,20,60,192,190,184,215,166],"version":0,"chains":[42,0],"bondTier":0}
2025-03-24 19:24:03.848 [TRC] SRV[127.0.0.1:50953]: Tatanka node handling message from remote tatanka. route = tatanka_config, payload = {"id":[2,141,62,254,88,87,165,131,30,223,95,103,198,85,131,185,163,98,70,208,34,127,0,79,126,174,68,10,54,54,39,180,143],"version":0,"chains":[42,0],"bondTier":0}
2025-03-24 19:24:04.827 [TRC] tCL[0:127.0.0.1:50952][tTC][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:24:04.833 [DBG] COMMS: Starting websocket handler for 127.0.0.1:50962
2025-03-24 19:24:04.833 [TRC] COMMS: New websocket client 127.0.0.1
2025-03-24 19:24:04.833 [TRC] COMMS[WS]: Starting websocket messaging with peer 127.0.0.1
2025-03-24 19:24:04.840 [TRC] SRV[127.0.0.1:50952]: Tatanka node handling message. route = post_bond, payload = [{"peerID":[2,172,50,212,29,142,218,235,196,186,192,103,137,144,32,183,111,251,62,58,201,101,174,44,184,202,1,83,243,177,7,71,67],"assetID":42,"coinID":"","strength":1,"expiration":"2026-03-24T19:24:04.833931+01:00","maturation":"0001-01-01T00:00:00Z"}]
2025-03-24 19:24:04.840 [ERR] SRV[127.0.0.1:50952]: Error storing bond for client 02ac32d41d8edaebc4bac067899020b76ffb3e3ac965ae2cb8ca0153f3b1074743 in db: no zero-length keys allowed
2025-03-24 19:24:04.841 [INF] tCL[0:127.0.0.1:50952][tTC][tTC]: Disconnecting old tatanka node 028d3efe5857a5831edf5f67c65583b9a36246d0227f004f7eae440a363627b48f @ ws://127.0.0.1:50952
2025-03-24 19:24:04.841 [TRC] tCL[0:127.0.0.1:50952][tTC][DB][BADG]: Storing value log head: {Fid:0 Len:42 Offset:639}

```

But this particular run did not crash.
```
2025-03-24 19:25:22.867 [TRC] SRV[127.0.0.1:51002][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:25:22.877 [INF] SRV[127.0.0.1:51002]: Starting Tatanka node with peer ID 03d61005ef3012c8898b3e069d17e2e6a468dc79c210600c4281b306bfeb8d5d58
2025-03-24 19:25:22.878 [INF] COMMS: Server listening on 127.0.0.1:51002
2025-03-24 19:25:22.878 [INF] SRV[127.0.0.1:51002]: Fiat rate source "CryptoCompare" is disabled...
2025-03-24 19:25:22.879 [ERR] SRV[127.0.0.1:51002]: error connecting boot node with proto = 02ae2e0c945c7dd9edde22ff96938eb5b450f949b62ec78e2347433e052ed8b966, config = {"url":"ws://127.0.0.1:51003","cert":null}
2025-03-24 19:25:23.871 [TRC] SRV[127.0.0.1:51003][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:25:23.884 [INF] SRV[127.0.0.1:51003]: Starting Tatanka node with peer ID 02ae2e0c945c7dd9edde22ff96938eb5b450f949b62ec78e2347433e052ed8b966
2025-03-24 19:25:23.884 [INF] COMMS: Server listening on 127.0.0.1:51003
2025-03-24 19:25:23.884 [INF] SRV[127.0.0.1:51003]: Fiat rate source "CryptoCompare" is disabled...
2025-03-24 19:25:23.885 [DBG] COMMS: Starting websocket handler for 127.0.0.1:51007
2025-03-24 19:25:23.885 [TRC] COMMS: New websocket client 127.0.0.1
2025-03-24 19:25:23.885 [TRC] COMMS[WS]: Starting websocket messaging with peer 127.0.0.1
2025-03-24 19:25:23.885 [INF] SRV[127.0.0.1:51003]: Connected to boot node with peer ID 03d61005ef3012c8898b3e069d17e2e6a468dc79c210600c4281b306bfeb8d5d58, config {"url":"ws://127.0.0.1:51002","cert":null}
2025-03-24 19:25:23.886 [TRC] SRV[127.0.0.1:51002]: Tatanka node handling message. route = tatanka_connect, payload = {"id":[2,174,46,12,148,92,125,217,237,222,34,255,150,147,142,181,180,80,249,73,182,46,199,142,35,71,67,62,5,46,216,185,102],"version":0,"chains":[42,0],"bondTier":0}
2025-03-24 19:25:23.891 [TRC] SRV[127.0.0.1:51003]: Tatanka node handling message from remote tatanka. route = tatanka_config, payload = {"id":[3,214,16,5,239,48,18,200,137,139,62,6,157,23,226,230,164,104,220,121,194,16,96,12,66,129,179,6,191,235,141,93,88],"version":0,"chains":[42,0],"bondTier":0}
2025-03-24 19:25:24.872 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: All 0 tables opened in 0s

2025-03-24 19:25:24.885 [DBG] COMMS: Starting websocket handler for 127.0.0.1:51010
2025-03-24 19:25:24.885 [TRC] COMMS: New websocket client 127.0.0.1
2025-03-24 19:25:24.885 [TRC] COMMS[WS]: Starting websocket messaging with peer 127.0.0.1
2025-03-24 19:25:24.887 [TRC] SRV[127.0.0.1:51002]: Tatanka node handling message. route = post_bond, payload = [{"peerID":[2,139,140,50,78,248,43,49,140,169,199,196,220,36,165,17,253,38,152,202,157,107,121,93,253,64,127,124,230,155,148,38,241],"assetID":42,"coinID":"","strength":1,"expiration":"2026-03-24T19:25:24.885909+01:00","maturation":"0001-01-01T00:00:00Z"}]
2025-03-24 19:25:24.887 [ERR] SRV[127.0.0.1:51002]: Error storing bond for client 028b8c324ef82b318ca9c7c4dc24a511fd2698ca9d6b795dfd407f7ce69b9426f1 in db: no zero-length keys allowed
2025-03-24 19:25:24.887 [INF] tCL[0:127.0.0.1:51002][tTC][tTC][TCP]: WebSockets client for ws://127.0.0.1:51002/ws has disconnected
2025-03-24 19:25:24.887 [DBG] tCL[0:127.0.0.1:51002][tTC][tTC][TCP][TC]: Sending close 1000 (normal) message.
2025-03-24 19:25:24.887 [INF] tCL[0:127.0.0.1:51002][tTC][tTC]: Disconnecting old tatanka node 03d61005ef3012c8898b3e069d17e2e6a468dc79c210600c4281b306bfeb8d5d58 @ ws://127.0.0.1:51002
2025-03-24 19:25:24.888 [TRC] COMMS[WS]: Sent 1 and dropped 0 messages to 127.0.0.1 before shutdown.
2025-03-24 19:25:24.888 [TRC] COMMS: Disconnected websocket client 127.0.0.1
2025-03-24 19:25:24.888 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Storing value log head: {Fid:0 Len:42 Offset:727}

2025-03-24 19:25:24.897 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Got compaction priority: {level:0 score:1.73 dropPrefixes:[]}
2025-03-24 19:25:24.897 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Running for level: 0

2025-03-24 19:25:24.898 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: LOG Compact. Added 8 keys. Skipped 0 keys. Iteration took: 1.603934ms
2025-03-24 19:25:24.903 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Discard stats: map[]
2025-03-24 19:25:24.909 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: LOG Compact 0->1, del 1 tables, add 1 tables, took 12.490572ms

2025-03-24 19:25:24.909 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Compaction for level: 0 DONE
2025-03-24 19:25:24.911 [TRC] tCL[0:127.0.0.1:51002][tTC][DB][BADG]: Force compaction on level 0 done
2025-03-24 19:25:24.920 [ERR] SRV[127.0.0.1:51003]: failed to retrieve rate from KuCoin: unable to fetch fiat rates: error performing request: Get "https://api.kucoin.com/api/v1/prices?currencies=DCR,BTC": context canceled
2025-03-24 19:25:24.920 [INF] COMMS: Server shutting down...
2025-03-24 19:25:24.920 [ERR] SRV[127.0.0.1:51002]: failed to retrieve rate from KuCoin: unable to fetch fiat rates: error performing request: Get "https://api.kucoin.com/api/v1/prices?currencies=DCR,BTC": context canceled
2025-03-24 19:25:24.920 [DBG] COMMS: RPC listener done for 127.0.0.1:51002
2025-03-24 19:25:24.920 [DBG] CL[ws://127.0.0.1:51002]: Sending close 1000 (normal) message.
2025-03-24 19:25:24.920 [ERR] SRV[127.0.0.1:51003]: failed to retrieve rate from Binance: unable to fetch fiat rates: error performing request: Get "https://api3.binance.com/api/v3/ticker/price?symbols=[%22DCRUSDT%22%2C%22BTCUSDT%22]": context canceled
2025-03-24 19:25:24.920 [ERR] SRV[127.0.0.1:51002]: failed to retrieve rate from Binance: unable to fetch fiat rates: error performing request: Get "https://api3.binance.com/api/v3/ticker/price?symbols=[%22DCRUSDT%22%2C%22BTCUSDT%22]": context canceled
2025-03-24 19:25:24.921 [ERR] SRV[127.0.0.1:51002]: failed to retrieve rate from Coinparika: error performing request: Get "https://api.coinpaprika.com/v1/tickers": context canceled
2025-03-24 19:25:24.920 [ERR] SRV[127.0.0.1:51003]: failed to retrieve rate from Coinparika: error performing request: Get "https://api.coinpaprika.com/v1/tickers": context canceled
2025-03-24 19:25:24.921 [INF] SRV[127.0.0.1:51003]: Fiat rate source "Messari" is disabled...
2025-03-24 19:25:24.920 [TRC] COMMS[WS]: Sent 2 and dropped 0 messages to 127.0.0.1 before shutdown.
2025-03-24 19:25:24.920 [INF] COMMS: Server shutting down...
2025-03-24 19:25:24.921 [TRC] COMMS: Disconnected websocket client 127.0.0.1
2025-03-24 19:25:24.921 [INF] SRV[127.0.0.1:51002]: Fiat rate source "Messari" is disabled...
error making first connected client: PostBond error: rpc error: error code 2: internal error

```

I'm thinking maybe my network connection was screwed whilst testing but maybe something else causing the panic.

I put a sleep in and it will panic every time, must be it.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x993702]

goroutine 275 [running]:
github.com/dgraph-io/badger/skl.(*Skiplist).IncrRef(...)
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/skl/skl.go:86
github.com/dgraph-io/badger.(*DB).getMemTables(0xc0000cb508)
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:493 +0xc2
github.com/dgraph-io/badger.(*DB).get(0xc0000cb508, {0xc006572e10, 0xb, 0xb})
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:525 +0x69
github.com/dgraph-io/badger.(*Txn).Get(0xc0000fb400, {0xc00011eb10, 0x3, 0x3})
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/txn.go:407 +0x287
github.com/dgraph-io/badger.(*Sequence).Release.func1(0xc0000fb400)
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:1131 +0x31
github.com/dgraph-io/badger.(*DB).Update(0x47e765?, 0xc00038d730)
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/txn.go:696 +0x74
github.com/dgraph-io/badger.(*Sequence).Release(0xc005e9a540)
        /home/joe/go/pkg/mod/github.com/dgraph-io/badger@v1.6.2/db.go:1130 +0x8b
decred.org/dcrdex/dex/lexi.(*DB).Connect.func1()
        /home/joe/git/dcrdex/dex/lexi/lexi.go:85 +0x7d
created by decred.org/dcrdex/dex/lexi.(*DB).Connect in goroutine 1
        /home/joe/git/dcrdex/dex/lexi/lexi.go:81 +0x90
```

--- a/dex/lexi/lexi.go
+++ b/dex/lexi/lexi.go
@@ -79,20 +79,16 @@ func New(cfg *Config) (*DB, error) {
 func (db *DB) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	db.wg.Add(1)
 	go func() {
-		defer db.wg.Done()
-		<-ctx.Done()
-		if err := db.idSeq.Release(); err != nil {
-			db.log.Errorf("Error releasing sequence: %v", err)
-		}
-	}()
-
-	db.wg.Add(1)
-	go func() {
-		defer db.wg.Done()
-		defer db.Close()
-		defer db.updateWG.Wait()
 		ticker := time.NewTicker(5 * time.Minute)
-		defer ticker.Stop()
+		defer func() {
+			ticker.Stop()
+			db.updateWG.Wait()
+			if err := db.idSeq.Release(); err != nil {
+				db.log.Errorf("Error releasing sequence: %v", err)
+			}
+			db.Close()
+			db.wg.Done()
+		}()
 		for {
 			select {
 			case <-ticker.C:

